api/src/main/java/org/pmiops/workbench/api/CloudTaskEnvironmentsController.java

```diff
@@ -0,0 +1,190 @@
+package org.pmiops.workbench.api;
+
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import org.pmiops.workbench.db.model.DbWorkspace;
+import org.pmiops.workbench.exceptions.WorkbenchException;
+import org.pmiops.workbench.leonardo.LeonardoApiClient;
+import org.pmiops.workbench.leonardo.LeonardoStatusUtils;
+import org.pmiops.workbench.model.UserRole;
+import org.pmiops.workbench.workspaces.WorkspaceService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class CloudTaskEnvironmentsController implements CloudTaskEnvironmentsApiDelegate {
+  private static final Logger LOGGER =
+      Logger.getLogger(CloudTaskEnvironmentsController.class.getName());
+
+  private final LeonardoApiClient leonardoApiClient;
+  private final WorkspaceService workspaceService;
+
+  @Autowired
+  public CloudTaskEnvironmentsController(
+      LeonardoApiClient leonardoApiClient, WorkspaceService workspaceService) {
+    this.leonardoApiClient = leonardoApiClient;
+    this.workspaceService = workspaceService;
+  }
+
+  @Override
+  public ResponseEntity<Void> ctDeleteUnsharedWorkspaceEnvironments(
+      List<String> workspaceNamespaces) {
+    LOGGER.info(
+        String.format(
+            "Deleting unshared environments for %d workspaces.", workspaceNamespaces.size()));
+
+    workspaceNamespaces.stream()
+        .map(workspaceService::lookupWorkspaceByNamespace)
+        .forEach(
+            ws -> {
+              //              explore(ws);
+              deleteUnshared(ws);
+            });
+
+    return ResponseEntity.ok().build();
+  }
+
+  private void explore(DbWorkspace dbWorkspace) {
+    LOGGER.info(
+        String.format(
+            "Exploring workspace %s/%s (ID %d)",
+            dbWorkspace.getWorkspaceNamespace(),
+            dbWorkspace.getFirecloudName(),
+            dbWorkspace.getWorkspaceId()));
+
+    var roles =
+        workspaceService.getFirecloudUserRoles(
+            dbWorkspace.getWorkspaceNamespace(), dbWorkspace.getFirecloudName());
+    roles.forEach(
```

Comment:
TODO: should we remove environments when collabs have been demoted to READER status?  The current version does not do this.